### PR TITLE
[MNG-7553] Introduce SettingsBuilder and ToolchainsBuilder services

### PR DESCRIPTION
**Pull request by gnodet**
Adds `ToolchainsBuilder` and `SettingsBuilder` services.

### CODE REVIEW

**Comment by michael-o** [comment]
Can you elaborate on the change? When was it broken?

**Comment by gnodet** [comment]
> Can you elaborate on the change? When was it broken?

When the v4 api was merged.  A few services were migrated to the new API, but that was a mistake because they are in use by plugins.  This is the case for the `ToolchainsBuilder` and the `MojoDescriptorCreator` as reported on https://lists.apache.org/list.html?dev@maven.apache.org. Whether those are considered internal or part of the real api could be discussed, but they were broken, as now is not the time to break compatibility.
This PR then restores those services and add 2 new services to the new API to build settings and toolchains.

I need to investigate why the IT is failing.

**Review by cstamas — APPROVED** [review]

File: api/maven-api-core/src/main/java/org/apache/maven/api/services/ProjectBuilderResult.java [hunk 1]
```diff
@@ -70,7 +70,7 @@
      * @return The problems that were encountered during the project building, can be empty but never {@code null}.
      */
     @Nonnull
-    Collection<ProjectBuilderProblem> getProblems();
+    Collection<BuilderProblem> getProblems();
```
Comment (michael-o) [inline api/maven-api-core/src/main/java/org/apache/maven/api/services/ProjectBuilderResult.java:73]: Stupid question: Is it really a problem with the builder or a problem with the build?

File: api/maven-api-core/src/main/java/org/apache/maven/api/services/SettingsBuilder.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+    /**
+     * Builds the effective settings of the specified settings files.
+     *
+     * @param request The settings building request that holds the parameters, must not be {@code null}.
+     * @return The result of the settings building, never {@code null}.
+     * @throws SettingsBuilderException If the effective settings could not be built.
+     */
+    @Nonnull
+    SettingsBuilderResult build( @Nonnull SettingsBuilderRequest request );
```
Comment (michael-o) [inline api/maven-api-core/src/main/java/org/apache/maven/api/services/SettingsBuilder.java:42]: Same question here.

File: api/maven-api-core/src/main/java/org/apache/maven/api/services/SettingsBuilderRequest.java [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+    @Nonnull
+    Session getSession();
+
+    /**
+     * Gets the global settings path.
+     *
+     * @return The global settings path or {@code null} if none.
+     */
+    @Nonnull
+    Optional<Path> getGlobalSettingsPath();
```
Comment (michael-o) [inline api/maven-api-core/src/main/java/org/apache/maven/api/services/SettingsBuilderRequest.java:51]: Javadoc does not match code

File: api/maven-api-core/src/main/java/org/apache/maven/api/services/SettingsBuilderRequest.java [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+    @Nonnull
+    Optional<Source> getGlobalSettingsSource();
+
+    /**
+     * Gets the user settings path.
+     *
+     * @return The user settings path or {@code null} if none.
+     */
+    @Nonnull
+    Optional<Path> getUserSettingsPath();
```
Comment (michael-o) [inline api/maven-api-core/src/main/java/org/apache/maven/api/services/SettingsBuilderRequest.java:67]: Javadoc does not match code

File: api/maven-api-core/src/main/java/org/apache/maven/api/services/SettingsBuilderRequest.java [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+    static SettingsBuilderRequest build( @Nonnull Session session,
+                                         @Nonnull Path globalSettingsPath,
+                                         @Nonnull Path userSettingsPath )
+    {
+        return builder()
+                .session( nonNull( session, "session can not be null" ) )
+                .globalSettingsPath( nonNull( globalSettingsPath, "globalSettingsPath can not be null" ) )
+                .userSettingsPath( nonNull( userSettingsPath, "userSettingsPath can not be null" ) )
+                .build();
+    }
```
Comment (michael-o) [inline api/maven-api-core/src/main/java/org/apache/maven/api/services/SettingsBuilderRequest.java:99]: cannot

File: api/maven-api-core/src/main/java/org/apache/maven/api/services/ToolchainsBuilderException.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+ */
+@Experimental
+public class ToolchainsBuilderException
+    extends MavenException
+{
+    /**
+     * @param message The message to give.
+     * @param e The {@link Exception}.
+     */
+    public ToolchainsBuilderException( String message, Exception e )
```
Comment (michael-o) [inline api/maven-api-core/src/main/java/org/apache/maven/api/services/ToolchainsBuilderException.java:37]: Same here

File: api/maven-api-core/src/main/java/org/apache/maven/api/services/ToolchainsBuilderRequest.java [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+    @Nonnull
+    Session getSession();
+
+    /**
+     * Gets the global Toolchains path.
+     *
+     * @return The global Toolchains path or {@code null} if none.
+     */
+    @Nonnull
+    Optional<Path> getGlobalToolchainsPath();
```
Comment (michael-o) [inline api/maven-api-core/src/main/java/org/apache/maven/api/services/ToolchainsBuilderRequest.java:43]: Javadoc does not match code

File: api/maven-api-core/src/main/java/org/apache/maven/api/services/ToolchainsBuilderRequest.java [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+
+    /**
+     * Gets the global Toolchains source.
+     *
+     * @return The global Toolchains source or {@code null} if none.
+     */
+    @Nonnull
+    Optional<Source> getGlobalToolchainsSource();
```
Comment (michael-o) [inline api/maven-api-core/src/main/java/org/apache/maven/api/services/ToolchainsBuilderRequest.java:51]: Javadoc does not match code

File: api/maven-api-core/src/main/java/org/apache/maven/api/services/ToolchainsBuilderRequest.java [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+
+    /**
+     * Gets the user Toolchains path.
+     *
+     * @return The user Toolchains path or {@code null} if none.
+     */
+    @Nonnull
+    Optional<Path> getUserToolchainsPath();
```
Comment (michael-o) [inline api/maven-api-core/src/main/java/org/apache/maven/api/services/ToolchainsBuilderRequest.java:59]: Javadoc does not match code

File: api/maven-api-core/src/main/java/org/apache/maven/api/services/ToolchainsBuilderRequest.java [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+
+    /**
+     * Gets the user Toolchains source.
+     *
+     * @return The user Toolchains source or {@code null} if none.
+     */
+    @Nonnull
+    Optional<Source> getUserToolchainsSource();
```
Comment (michael-o) [inline api/maven-api-core/src/main/java/org/apache/maven/api/services/ToolchainsBuilderRequest.java:67]: Javadoc does not match code

File: api/maven-api-core/src/main/java/org/apache/maven/api/services/ToolchainsBuilderRequest.java [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+    static ToolchainsBuilderRequest build( @Nonnull Session session,
+                                         @Nonnull Source globalToolchainsSource,
+                                         @Nonnull Source userToolchainsSource )
+    {
+        return builder()
+                .session( nonNull( session, "session can not be null" ) )
+                .globalToolchainsSource( nonNull( globalToolchainsSource, "globalToolchainsSource can not be null" ) )
+                .userToolchainsSource( nonNull( userToolchainsSource, "userToolchainsSource can not be null" ) )
+                .build();
+    }
```
Comment (michael-o) [inline api/maven-api-core/src/main/java/org/apache/maven/api/services/ToolchainsBuilderRequest.java:79]: cannot

File: api/maven-api-core/src/main/java/org/apache/maven/api/services/ToolchainsBuilderRequest.java [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+    static ToolchainsBuilderRequest build( @Nonnull Session session,
+                                         @Nonnull Path globalToolchainsPath,
+                                         @Nonnull Path userToolchainsPath )
+    {
+        return builder()
+                .session( nonNull( session, "session can not be null" ) )
+                .globalToolchainsPath( nonNull( globalToolchainsPath, "globalToolchainsPath can not be null" ) )
+                .userToolchainsPath( nonNull( userToolchainsPath, "userToolchainsPath can not be null" ) )
+                .build();
+    }
```
Comment (michael-o) [inline api/maven-api-core/src/main/java/org/apache/maven/api/services/ToolchainsBuilderRequest.java:91]: cannot

File: maven-core/src/main/java/org/apache/maven/internal/impl/DefaultProjectBuilder.java [hunk 13]
```diff
@@ -228,9 +228,9 @@ public String getMessage()
                         }
 
                         @Override
-                        public ProjectBuilderProblemSeverity getSeverity()
+                        public BuilderProblemSeverity getSeverity()
```
Comment (michael-o) [inline maven-core/src/main/java/org/apache/maven/internal/impl/DefaultProjectBuilder.java:231]: Why isn't it `BuilderProblem.Severity`?

**Review by michael-o — CHANGES_REQUESTED** [review]
I see two general issues/questions here:
* We have already briefly discussed the benefit/overrating of `Optional`. Can you explain why it makes sense here? I mean `X != null` is as good as here.
* I have a general problem with the terms `XBuilderRequest`/`XBuilderResult`/`XBuilderException`. I think this is incorrectly names since it implies that there is something with the builder and not the build itself. In other spots we use the terms `XBuildingRequest`/`XBuildingResult`/`XBuildingException`/`XBuildingProblem` or just `Build`. @elharo

File: api/maven-api-core/src/main/java/org/apache/maven/api/Session.java [hunk 14]
```diff
@@ -56,9 +56,21 @@
     @Nonnull
     SessionData getData();
 
+    /**
+     * Gets the user properties to use for interpolation. The user properties have been configured directly by the user
+     * on his discretion, e.g. via the {@code -Dkey=value} parameter on the command line.
```
Comment (elharo) [inline api/maven-api-core/src/main/java/org/apache/maven/api/Session.java:61]: delete "on his discretion"
, --> ;

File: api/maven-api-core/src/main/java/org/apache/maven/api/Session.java [hunk 15]
```diff
@@ -56,9 +56,21 @@
...
+     *
+     * @return The user properties, never {@code null}.
+     */
     @Nonnull
     Map<String, String> getUserProperties();
 
+    /**
+     * Gets the system properties to use for interpolation. The system properties are collected from the runtime
+     * environment like {@link System#getProperties()} and environment variables.
```
Comment (elharo) [inline api/maven-api-core/src/main/java/org/apache/maven/api/Session.java:70]: like --> such as

File: api/maven-api-core/src/main/java/org/apache/maven/api/Session.java [hunk 16]
```diff
@@ -56,9 +56,21 @@
     @Nonnull
     SessionData getData();
 
+    /**
```
Comment (elharo) [inline api/maven-api-core/src/main/java/org/apache/maven/api/Session.java:59]: Javadoc additions to existing API are useful but can be included in a separate PR that is easier to approve.

File: api/maven-api-core/src/main/java/org/apache/maven/api/services/BuilderProblemSeverity.java [hunk 17]
```diff
@@ -27,7 +27,7 @@
  * @since 4.0
  */
 @Experimental
-public enum ProjectBuilderProblemSeverity
+public enum BuilderProblemSeverity
```
Comment (elharo) [inline api/maven-api-core/src/main/java/org/apache/maven/api/services/BuilderProblemSeverity.java:30]: rename seems unnecessary

File: api/maven-api-core/src/main/java/org/apache/maven/api/services/ProjectBuilder.java [hunk 18]
```diff
@@ -51,13 +51,13 @@ public interface ProjectBuilder extends Service
      * Creates a {@link org.apache.maven.api.Project} from a POM file.
      *
      * @param session The {@link Session}, must not be {@code null}.
-     * @param source The {@link ProjectBuilderSource}, must not be {@code null}.
-     * @throws ProjectBuilderException if the project cannot be created
+     * @param source The {@link Source}, must not be {@code null}.
```
Comment (elharo) [inline api/maven-api-core/src/main/java/org/apache/maven/api/services/ProjectBuilder.java:54]: The --> the

File: api/maven-api-core/src/main/java/org/apache/maven/api/services/ProjectBuilder.java [hunk 19]
```diff
@@ -51,13 +51,13 @@ public interface ProjectBuilder extends Service
...
+     * @throws ProjectBuilderException if the project can not be created
```
Comment (elharo) [inline api/maven-api-core/src/main/java/org/apache/maven/api/services/ProjectBuilder.java:55]: revert, cannot is one word

File: api/maven-api-core/src/main/java/org/apache/maven/api/services/SettingsBuilder.java [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+ */
+
+import java.nio.file.Path;
+
+import org.apache.maven.api.Service;
+import org.apache.maven.api.Session;
+import org.apache.maven.api.annotations.Nonnull;
+
+/**
+ * Builds the effective settings from a user settings file and/or a global settings file.
```
Comment (elharo) [inline api/maven-api-core/src/main/java/org/apache/maven/api/services/SettingsBuilder.java:29]: what is "effective settings"? How does that differ from other settings?

File: api/maven-api-core/src/main/java/org/apache/maven/api/services/SettingsBuilder.java [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+ */
+public interface SettingsBuilder extends Service
+{
+
+    /**
+     * Builds the effective settings of the specified settings files.
+     *
+     * @param request The settings building request that holds the parameters, must not be {@code null}.
+     * @return The result of the settings building, never {@code null}.
+     * @throws SettingsBuilderException If the effective settings could not be built.
```
Comment (elharo) [inline api/maven-api-core/src/main/java/org/apache/maven/api/services/SettingsBuilder.java:39]: if the effective settings could not be built

File: api/maven-api-core/src/main/java/org/apache/maven/api/services/SettingsBuilder.java [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+/**
+ * Builds the effective settings from a user settings file and/or a global settings file.
+ */
+public interface SettingsBuilder extends Service
+{
+
+    /**
+     * Builds the effective settings of the specified settings files.
+     *
+     * @param request The settings building request that holds the parameters, must not be {@code null}.
```
Comment (elharo) [inline api/maven-api-core/src/main/java/org/apache/maven/api/services/SettingsBuilder.java:37]: The --> the (per Oracle, API doc params do not begin with a capital letter

File: api/maven-api-core/src/main/java/org/apache/maven/api/services/SettingsBuilderException.java [hunk 23]
```diff
@@ -0,0 +1,43 @@
...
+ *
+ * @since 4.0
+ */
+@Experimental
+public class SettingsBuilderException
+    extends MavenException
+{
+    /**
+     * @param message The message to give.
+     * @param e The {@link Exception}.
```
Comment (elharo) [inline api/maven-api-core/src/main/java/org/apache/maven/api/services/SettingsBuilderException.java:35]: don't repeat type information in doc comments. It's redundant. Instead explain what this exception is

**Review by elharo — COMMENTED** [review]
I haven't read the whole thin yet, but I tend to agree that Optional is at best as bad as null, and sometimes worse.

**Comment by gnodet** [comment]
I'm going to split that PR in 2.  This one will be kept for the modifications on the v4 api : the new services and the related discussions.
I'll create a separate PR for the restoration of the API.

**Review by michael-o — APPROVED** [review]
Clean and simple

**Comment by gnodet** [comment]
> I see two general issues/questions here:
I would rather have had those discussions when I asked for feedback on the API, but I suppose it's not too late ;-)
See my answers / reasoning below.

> 
> * We have already briefly discussed the benefit/overrating of `Optional`. Can you explain why it makes sense here? I mean `X != null` is as good as here.

I slightly disagree.  The main differences are that when using an `Optional`, the code carries the semantics and actually forces the caller to deal with it, whereas just if you simply return a value (the optionals are only used as return values), you have to get to the definition, read the javadoc, so that you can know if you need to deal with a possible null value.  There are very often places in our code where you see both null checks and not.

This also hurts code concision.  A simple example found in the code:
```
        if ( getBaseVersionInternal() != null )
        {
            sb.append( getBaseVersionInternal() );
        }
        else
        {
            sb.append( versionRange.toString() );
        }
```
This could be rewritten as:
```
       sb.append( getBaseVersionInternal().orElse( versionRange.toString() ) );
```
This is actually related to the fact that the maven coding style is very sparse. I actually find that quite difficult to work with (even if I'm getting used to it).  That may be due to the fact I'm working almost exclusively on a laptop, so the height of it is limited, and only seing 30-40 lines of code at the same time becomes a problem when nearly 50% of those don't really convey any semantic (being blank lines or braces usually).  It means more time to scroll through the code back and forth to understand it.  I don't like that. Streams are another way to get around empty lines, just because often, you simply don't need braces when using them, so you have 1 line instead of 3.

So generally speaking, the more semantic carried by the code, the better.  I agree to not overuse `Optional`, I just don't think they are at this point.

> * I have a general problem with the terms `XBuilderRequest`/`XBuilderResult`/`XBuilderException`. I think this is incorrectly names since it implies that there is something with the builder and not the build itself. In other spots we use the terms `XBuildingRequest`/`XBuildingResult`/`XBuildingException`/`XBuildingProblem` or just `Build`. @elharo Sample: https://github.com/apache/maven/blob/35b93b0a589752cc88105623a2ddf9e52b56c1ce/maven-model-builder/src/main/java/org/apache/maven/model/building/DefaultModelBuildingRequest.java

I'm actually seeing the terms as _a request to the X builder_, _the result returned by X builder_, _the exception thrown by X builder_.  We can't use quotes, but think about those as _XBuilder's request_, _XBuilder's result_, _XBuilder's exception_.  I actually think those make more sense than what we have in other places.  It's also easier to see the relationship between the service and its related request/result/exception.  
Also, if you think each service usually throws _its own exception_. If we move the semantic to _what action is performed_ instead of _who does the action_, the exceptions would have to be more shared between services.  However (and that's still missing right now), a bunch of exception are used to carry the _context_ in which the exception was thrown.  So you end up with untyped `source` fields or similar things.

File: maven-core/src/main/java/org/apache/maven/internal/impl/DefaultProjectBuilder.java (hunk 13 above)
Comment (gnodet) [reply maven-core/src/main/java/org/apache/maven/internal/impl/DefaultProjectBuilder.java:231]: No reason, I'll move it as an inner enum.

File: api/maven-api-core/src/main/java/org/apache/maven/api/services/SettingsBuilder.java (hunk 20 above)
Comment (elharo) [reply api/maven-api-core/src/main/java/org/apache/maven/api/services/SettingsBuilder.java:29]: You need to put this in the docs, perhaps rewriting without the word effective

**Comment by michael-o** [comment]
> > I see two general issues/questions here:
> > I would rather have had those discussions when I asked for feedback on the API, but I suppose it's not too late ;-)
> > See my answers / reasoning below.
> 
> > * We have already briefly discussed the benefit/overrating of `Optional`. Can you explain why it makes sense here? I mean `X != null` is as good as here.
> 
> I slightly disagree. The main differences are that when using an `Optional`, the code carries the semantics and actually forces the caller to deal with it, whereas just if you simply return a value (the optionals are only used as return values), you have to get to the definition, read the javadoc, so that you can know if you need to deal with a possible null value. There are very often places in our code where you see both null checks and not and people end up adding null checks, just in case.
> 
> This also hurts code concision. A simple example found in the code:
> 
> ```
>         if ( getBaseVersionInternal() != null )
>         {
>             sb.append( getBaseVersionInternal() );
>         }
>         else
>         {
>             sb.append( versionRange.toString() );
>         }
> ```
> 
> This could be rewritten as:
> 
> ```
>        sb.append( getBaseVersionInternal().orElse( versionRange.toString() ) );
> ```
> 
> This is actually related to the fact that the maven coding style is very sparse. I actually find that quite difficult to work with (even if I'm getting used to it). That may be due to the fact I'm working almost exclusively on a laptop, so the height of it is limited, and only seing 30-40 lines of code at the same time becomes a problem when nearly 50% of those don't really convey any semantic (being blank lines or braces usually). It means more time to scroll through the code back and forth to understand it. I don't like that. Streams are another way to get around empty lines, just because often, you simply don't need braces when using them, so you have 1 line instead of 3.
> 
> So generally speaking, the more semantic carried by the code, the better. I agree to not overuse `Optional`, I just don't think they are at this point. That said, the use of `Nullable` / `NonNull` annotations does help with semantics already, and I would not be opposed to removing them on requests / results objects.

This reasoning I can follow where `Optional` adds value. I don't like like to proclaim `Optional` to be the new `null` which is non sense. No objections here.

> > * I have a general problem with the terms `XBuilderRequest`/`XBuilderResult`/`XBuilderException`. I think this is incorrectly names since it implies that there is something with the builder and not the build itself. In other spots we use the terms `XBuildingRequest`/`XBuildingResult`/`XBuildingException`/`XBuildingProblem` or just `Build`. @elharo Sample: https://github.com/apache/maven/blob/35b93b0a589752cc88105623a2ddf9e52b56c1ce/maven-model-builder/src/main/java/org/apache/maven/model/building/DefaultModelBuildingRequest.java
> 
> I'm actually seeing the terms as _a request to the X builder_, _the result returned by X builder_, _the exception thrown by X builder_. We can't use quotes, but think about those as _XBuilder's request_, _XBuilder's result_, _XBuilder's exception_. I actually think those make more sense than what we have in other places. It's also easier to see the relationship between the service and its related request/result/exception. Also, if you think each service usually throws _its own exception_. If we move the semantic to _what action is performed_ instead of _who does the action_, the exceptions would have to be more shared between services. However (and that's still missing on some right now), a bunch of exception are used to carry the _context_ in which the exception was thrown.
> 
> Another point that just came to my mind is that wording like `DependencyCollectionRequest` is not intuitive either: the term `Collection` in java is immediately associated to a... `java.util.Collection`, and not to the act of collecting...

Let me crunch on this. @elharo What is your opinion on the style here?

**Comment by gnodet** [comment]
This PR has been split with #852, #853, #854 for ease of review.
